### PR TITLE
filter out pods that have deletion timestamp set in currentlyDrainedPods

### LINKED ISSUE #6718: Cluster Autoscaler tries to schedule evicted pods on drained nodes
**Which component are you using?**:
cluster-autoscaler

**What version of the component are you using?**:
1.26.3

**What k8s version are you using (`kubectl version`)?**:
1.26

**What environment is this in?**:
EKS

**What did you expect to happen?**:
cluster autoscaler shouldn't try to simulate the scheduling behavior for pods that are evicted (api-server return 201 for the /evcition request) on the drained nodes.

**What happened instead?**:

during the first cluster-autocaler reconcilation, cluster-autoscaler decided to scale down a underutilized node (our threashold is 70)
```
I0414 10:21:40.188957 hinting_simulator.go:77] Pod pod-hbhqf can be moved to ip-10-160-48-58.us-west-2.compute.internal
I0414 10:21:40.188962 hinting_simulator.go:77] Pod pod-lx6fg can be moved to ip-10-160-52-115.us-west-2.compute.internal
I0414 10:21:40.293272       1 actuator.go:212] Scale-down: removing node ip-10-160-50-114.us-west-2.compute.internal, utilization: {0.6940581542351454 0.6892291641475079 0 cpu 0.6940581542351454}, pods to reschedule: pod-hbhqf,pod-lx6fg
```
the node drain started, and we have PDB maxUnavailable to set 1, in the audit log, we observed `pod-hbhqf ` eviction call succeeded, the new pod `pod-r9t79` coming up and got scheduled (but not ready yet), and `pod-lx6fg` eviction call failed. 
```
- 03:21:45.446 AM
   PDB status disruptionsAllowed: 0
   Kubelet killing pod-hbhqf container app
- 03:21:45.474 AM
   /api/v1/namespaces/ns/pods/pod-hbhqf/eviction returned 201
- 03:21:45.697 AM
   pod-r9t79 pod creation returned 201
- 03:21:45.698 AM
  Bind pod-r9t79 to node ip-10-160-52-115.us-west-2.compute.internal succeeded
- 03:21:45.975 AM
  /api/v1/namespaces/ns/pods/pod-lx6fg/eviction returned 429 (PDB violation)
```
during the second cluster-autoscaler reconcilation, it tried to bring up a new node in the AWS Autoscale scaling group:
```
I0414 10:21:50.361899 hinting_simulator.go:77] Pod pod-hbhqf can be moved to ip-10-160-48-58.us-west-2.compute.internal
I0414 10:21:50.362342 hinting_simulator.go:116] failed to find place for pod-lx6fg: cannot put pod-lx6fg on any node
I0414 10:21:50.485239       1 event_sink_logging_wrapper.go:48] Event(v1.ObjectReference{Kind:"Pod", Namespace:"ns", Name:"pod-lx6fg", APIVersion:"v1", FieldPath:""}): type: 'Normal' reason: 'TriggeredScaleUp' pod triggered scale-up: [{asg-name 9->10 (max: 10)}]
```

**Analysis**
We think that the current code puts all pods on the draining nodes to the unscheduable pod list, then runs the scheduling simulation. But actually some of the pods are the draining nodes are already in the eviction process and the deletionTimestamp have been set.
Should we ignore the pods that the eviction are allowed?
https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/#how-api-initiated-eviction-works

### PR DESCRIPTION
**Pull request by ZihanJiang96**
#### What type of PR is this?
/kind bug
/kind regression

#### What this PR does / why we need it:
Shouldn't put pods that have deletionTimestamp set on the draining node to the unshedulable pods list.
The issue has more detailed analysis. Thanks!!!

#### Which issue(s) this PR fixes:
Fixes #6718

### CODE REVIEW

**Comment by ZihanJiang96** [comment]
Hi @BigDarkClown @x13n @mwielgus could you please kindly review this PR? Thanks!

**Comment by x13n** [comment]
Thanks, the change makes sense to me. Can you update the unit tests to handle this scenario?

/ok-to-test

**Comment by ZihanJiang96** [comment]
"/easycla"

**Comment by ZihanJiang96** [comment]
thanks @x13n added unit test

File: cluster-autoscaler/core/podlistprocessor/currently_drained_nodes.go [hunk 1]
```diff
@@ -51,6 +51,11 @@ func currentlyDrainedPods(context *context.AutoscalingContext) []*apiv1.Pod {
 			continue
 		}
 		for _, podInfo := range nodeInfo.Pods {
+			// Filter out pods that has deletion timestamp set
+			if podInfo.Pod.DeletionTimestamp != nil {
+				klog.Infof("Pod %v has deletion timestamp set, skipping", podInfo.Pod.Name)
```
Comment (x13n) [inline cluster-autoscaler/core/podlistprocessor/currently_drained_nodes.go:56]: nit: I'd slightly rephrase the log line to make it a bit clearer to anyone not  familiar with this specific part of the code. From the logs, it won't be clear what exactly is skipped for the pod. `skipping injection into the pending pod list`?

**Review by x13n — CHANGES_REQUESTED** [review]

File: cluster-autoscaler/core/podlistprocessor/currently_drained_nodes.go (hunk 1 above)
Comment (ZihanJiang96) [reply cluster-autoscaler/core/podlistprocessor/currently_drained_nodes.go]: done

File: cluster-autoscaler/utils/test/test_utils.go [hunk 2]
```diff
@@ -195,6 +195,13 @@ func BuildScheduledTestPod(name string, cpu, memory int64, nodeName string) *api
 	return p
 }
 
+// BuildTestPodWithDeletionTimestamp builds a test pod with deletion timestamp set
+func BuildTestPodWithDeletionTimestamp(name string, cpu, memory int64, nodeName string, deletionTimestamp time.Time) *apiv1.Pod {
```
[the remark this answers is not in the document]
Comment (ZihanJiang96) [reply cluster-autoscaler/utils/test/test_utils.go]: done

**Comment by x13n** [comment]
Thanks!

/lgtm
/approve

**Pull request merged**